### PR TITLE
Add ability to set GCE node labels via Salt Cloud

### LINKED ISSUE #61245: [FEATURE REQUEST] Node Label management in Salt Cloud
**Is your feature request related to a problem? Please describe.**
I would like to request the ability to add labels to nodes provisioned by Salt Cloud (Node Labels)

**Describe the solution you'd like**
The following code block is what I would like to be made available to Salt Cloud profile configurations

```
labels:
  key: value
```

**Describe alternatives you've considered**
N/A

**Additional context**
Google Cloud is my primary baseline: https://cloud.google.com/compute/docs/labeling-resources

**Please Note**
If this feature request would be considered a substantial change or addition, this should go through a SEP process here https://github.com/saltstack/salt-enhancement-proposals, instead of a feature request.

### PR DESCRIPTION
**Pull request by sentient-glare**
### What does this PR do?
This PR adds the ability to set node labels via Salt Cloud in Google Compute Engine.

### What issues does this PR fix or reference?
Fixes: #61245

### New Behavior
See relevant issue for details.

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**

- [x] Docs
- [x] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
No

Please review [Salt's Contributing Guide](https://docs.saltproject.io/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

### CODE REVIEW

**Comment by sentient-glare** [comment]
First time contributor - I am a little lost when it comes to updating documentation and tests. Happy to update the MR if I can get a little more guidance.

Tests - I see there is a `tests/integration/cloud/clouds/test_gce.py` but as far as I can tell this just sees whether an instance can be created and destroyed. As far as I can tell there are currently no GCE unit tests in `tests/unit/`. Should I be writing new tests or just ensuring that the existing integration tests run, with labels set on the resulting instance?

Documentation - let me know if there's anything else I should be updating.

File: doc/topics/cloud/gce.rst [hunk 1]
```diff
@@ -235,6 +237,13 @@ Additionally, the subnetwork your instance is created under is associated with t
 
 .. versionadded:: 2017.7.0
 
+labels
+----
+
+This setting allows you to set labels on your GCE instances. It
+should be a dictionary and must be parse-able by the python
+ast.literal_eval() function to convert it to a python dictionary.
+
```
Comment (Ch3LL) [inline doc/topics/cloud/gce.rst:246]: can you add a versionadded tag here: `.. versionadded:: 3006`

**Review by Ch3LL — CHANGES_REQUESTED** [review]

**Comment by Ch3LL** [comment]
@waynew any ideas on how we could add better test coverage for this one?

File: doc/topics/cloud/gce.rst (hunk 1 above)
Comment (sentient-glare) [reply doc/topics/cloud/gce.rst:246]: done!

**Comment by waynew** [comment]
tl;dr - I think the code is probably correct but the unit tests for this module
(not just this PR) are pretty broken.

---

@Ch3LL / @sentient-glare - I went ahead and made some tweaks here on today's
Test Clinic - you should be able to catch the replay at
https://twitch.tv/saltprojectoss if you take a look in the next few days. I
also pushed my changes to
https://github.com/waynew/salt/pull/new/tc/gce-test-tweaks

I have some pretty significant concerns -- not so much about this PR which
looks good AFAICT -- but more like the test cases and also some huge questions
around if the config (for the module, not just this PR) is documented correctly
or what.

Basically, as I understand it, the `conn.create_node` call (line 2417
currently) wants `ex_labels` or `ex_tags` as keyword arguments. In the code
where we `get_cloud_config_value` we're looking for `labels`, `tags`,
`metadata`, etc.

However in the `config` fixture in the tests we're actually specifying the
config values as the kwargs (e.g. `ex_labels`, `ex_tags`, etc.) which uh...
aren't ever used, so `None` or whatever the default values are get passed
instead. The tests are only passing because the specified config values are
completely ignored.

In my branch/for the Test Clinic I went through how to modify the config to
require fixtures, as well as add parameters to the fixtures so we can make sure
that we're actually providing/checking for the correct values. For each of the
`def __get_XXX` config functions we should be providing:

1. Correct, parseable values (lists, dictionaries, whatever is required)
2. Incorrect, unparseable values (TBH, these functions should probably catch a
   `SyntaxError` instead of general `Exception`)
3. Incorrect, but parseable values: empty data structures, empty strings,
   incorrect datatypes, e.g. list where a string belongs. (We should also
   probably be logging a warning when we can't actually parse that data, too,
   since if it's present then someone would be expecting it to be parsed
   correctly and not silently just fail).

That interpretation/understanding may be incorrect -- I wasn't actually doing
anything with the real connection/library, and I didn't dig into the docs that
much. 

Personally I think that *correct* unit tests should be enough here -- I'm
pretty sure this is one of the modules that we'll want to extract to an
extension module "soon" (for varying sizes of soon) anyway.

@sentient-glare - if my PR/Test Clinic isn't enough to get these tests all
sorted out, feel free to reach out to me on IRC (waynew in #salt on
Libera.chat), or join Thursday's Test Clinic (or next Tuesday, or whenever you
can).

HTH!

**Comment by Ch3LL** [comment]
bump @sentient-glare  did you see @waynew 's comment above ^

**Comment by sentient-glare** [comment]
Seen! Been on vacation the last little while, looks like I missed the replay but I'll take a look at the branch mentioned.

**Comment by sentient-glare** [comment]
@waynew @Ch3LL I've gone through and given this an update - some of the __get_x are not currently included in the tests as they are not used in request_instance.

File: salt/cloud/clouds/gce.py [hunk 2]
```diff
@@ -369,9 +369,12 @@ def __get_image(conn, vm_):
     The get_image for GCE allows partial name matching and returns a
     libcloud object.
     """
+    default_img = "debian-7"
     img = config.get_cloud_config_value(
-        "image", vm_, __opts__, default="debian-7", search_global=False
+        "image", vm_, __opts__, default=default_img, search_global=False
     )
+    if not img or not isinstance(img, str):
```
Comment (Ch3LL) [inline salt/cloud/clouds/gce.py:376]: The `get_cloud_config_value` should return `debian-7` if it can't find image set anywhere. Can you elaborate why you are adding this if statement here?

File: salt/cloud/clouds/gce.py [hunk 3]
```diff
@@ -387,18 +390,39 @@ def __get_size(conn, vm_):
...
+    return labels
+
+
 def __get_tags(vm_):
     """
     Get configured tags.
     """
     t = config.get_cloud_config_value(
-        "tags", vm_, __opts__, default="[]", search_global=False
+        "ex_tags", vm_, __opts__, default="[]", search_global=False
```
Comment (Ch3LL) [inline salt/cloud/clouds/gce.py:425]: why does this need to be changed to `ex_`?

**Review by Ch3LL — COMMENTED** [review]
just a couple more questions

File: salt/cloud/clouds/gce.py (hunk 2 above)
Comment (sentient-glare) [reply salt/cloud/clouds/gce.py:376]: Added purely to get [this test case](https://github.com/saltstack/salt/pull/62046/files/5cd5fd75bb7eb7a60e6d73d4efc7611e482ebff9#diff-011c0910037d4cbee4d337f9311903470b81753065ce3586eff691d536f5a465R55) to pass. Could remove it and the case

File: salt/cloud/clouds/gce.py (hunk 3 above)
Comment (sentient-glare) [reply salt/cloud/clouds/gce.py:425]: Right - this was changed originally because the test config originally provided in test_gce.py provided `ex_tags`.
https://github.com/saltstack/salt/pull/62046/files/5cd5fd75bb7eb7a60e6d73d4efc7611e482ebff9#diff-011c0910037d4cbee4d337f9311903470b81753065ce3586eff691d536f5a465L69

This isn't correct though, `tags` the is the right key. Have put this back and corrected the test config

File: salt/cloud/clouds/gce.py [hunk 4]
```diff
@@ -416,7 +440,7 @@ def __get_metadata(vm_):
     Get configured metadata and add 'salt-cloud-profile'.
     """
     md = config.get_cloud_config_value(
-        "metadata", vm_, __opts__, default="{}", search_global=False
+        "ex_metadata", vm_, __opts__, default="{}", search_global=False
```
Comment (Ch3LL) [inline salt/cloud/clouds/gce.py:443]: sorry i missed this one in the original review. doesn't this also need to remain metadata and not `ex_metadata`?

File: salt/cloud/clouds/gce.py (hunk 2 above)
Comment (Ch3LL) [reply salt/cloud/clouds/gce.py:376]: I would prefer to remove the if statement and fix the test. We should be able to mock what is returned by `config.get_cloud_config_value`

File: salt/cloud/clouds/gce.py (hunk 4 above)
Comment (Ch3LL) [inline salt/cloud/clouds/gce.py:443]: sorry i missed this one in the original review. doesn't this also need to remain metadata and not ex_metadata?

**Review by Ch3LL — CHANGES_REQUESTED** [review]

File: salt/cloud/clouds/gce.py (hunk 4 above)
Comment (sentient-glare) [reply salt/cloud/clouds/gce.py]: Right again - fixed

File: salt/cloud/clouds/gce.py [hunk 5]
```diff
@@ -387,12 +387,33 @@ def __get_size(conn, vm_):
     """
     Need to override libcloud to find the machine type in the proper zone.
     """
+    default_size = "n1-standard-1"
     size = config.get_cloud_config_value(
-        "size", vm_, __opts__, default="n1-standard-1", search_global=False
+        "size", vm_, __opts__, default=default_size, search_global=False
     )
+    if not size or not isinstance(size, str):
```
Comment (Ch3LL) [inline salt/cloud/clouds/gce.py:394]: Can we remove this as discussed previously. `get_cloud_config_value` will return the default if its not set.

File: salt/cloud/clouds/gce.py [hunk 6]
```diff
@@ -456,9 +477,12 @@ def __get_network(conn, vm_):
     """
     Return a GCE libcloud network object with matching name
     """
+    default_network = "default"
     network = config.get_cloud_config_value(
-        "network", vm_, __opts__, default="default", search_global=False
+        "network", vm_, __opts__, default=default_network, search_global=False
     )
+    if not network or not isinstance(network, str):
```
Comment (Ch3LL) [inline salt/cloud/clouds/gce.py:484]: same here

**Review by Ch3LL — CHANGES_REQUESTED** [review]

**Review by Ch3LL — DISMISSED** [review]

**Comment by Ch3LL** [comment]
looks like there are test failures related to this change

**Review by Ch3LL — DISMISSED** [review]

**Comment by Ch3LL** [comment]
There is a doc failure:

```
 Warning, treated as error:
/__w/salt/salt/doc/topics/cloud/gce.rst:241:Title underline too short.

labels
----
[ERROR   ] Unable to load range library
```

**Review by garethgreenaway — APPROVED** [review]
